Subscribe to attestation committee index & test subscription

cemozerr opened the pull request:
<!-- Thanks for sending a pull request! Please check out our contribution guidelines: -->
<!-- https://github.com/PegaSysEng/artemis/blob/master/CONTRIBUTING.md -->

## PR Description
Implements logic for Validator Coordinator to registers & deregister attestation subnets by emitting events to GossipMessageHandler.  Subscribes to attestation subnets one epoch ahead, as the spec suggests.

cemozerr committed:
Subscribe to attestation committee index & test subscription

cemozerr committed:
Minor cleanup & spotless

cemozerr committed:
Change magic number to constant

cemozerr committed:
Implement deregistering of attestation subnets

ajsutton commented on ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/operations/AttestationData.java:
```diff
@@ -60,8 +60,8 @@ public AttestationData(UnsignedLong slot, AttestationData data) {
-    this.source = data.getSource();
-    this.target = data.getTarget();
+    this.source = new Checkpoint(data.getSource());
+    this.target = new Checkpoint(data.getTarget());
```
We could actually make `Checkpoint` immutable and not need to do this copy.  The setters are only used in a couple of tests which would could easily just use the `Checkpoint(UnsignedLong, Bytes32)` constructor instead.

cemozerr replied:
Agreed, but I'd rather do that in a separate PR: https://pegasys1.atlassian.net/secure/RapidBoard.jspa?rapidView=67&view=planning&selectedIssue=BC-168
I only made this small change here because I noticed that we have a potential bug here if we don't create a new Checkpoint object.

ajsutton commented on validator/coordinator/src/main/java/tech/pegasys/artemis/validator/coordinator/ValidatorCoordinator.java:
```diff
@@ -117,6 +124,8 @@
   private LinkedBlockingQueue<ProposerSlashing> slashings = new LinkedBlockingQueue<>();
   private HashMap<UnsignedLong, Eth1DataWithIndexAndDeposits> eth1DataCache = new HashMap<>();
 
+  @VisibleForTesting public Map<Integer, Integer> committeeIndexTTL = new HashMap<>();
```
I wouldn't make this public. If we really need it to be visible for testing add a `getCommitteeIndexTtl` method which can return an immutable view of the map. Otherwise we're exposing mutable state which the object should be encapsulating.

But at the moment there don't seem to be any tests using it so it could probably just be private.

ajsutton commented on validator/coordinator/src/main/java/tech/pegasys/artemis/validator/coordinator/ValidatorCoordinator.java:
```diff
@@ -253,7 +262,36 @@ public void onAttestationEvent(BroadcastAttestationEvent event) throws IllegalAr
+            committeeIndexTTL.remove(index);
+          } else {
+            entry.setValue(ttl);
+          }
```
This would be a lot simpler if instead of storing a map of TTLs we just stored the slot at which we should deregister the index.

So when registering:
`put(index, slot + SLOTS_PER_EPOCH * COMMITTEE_INDEX_SUBSCRIPTION_LENGTH)`

and then you remove and deregister if the current slot > slotInMap.

ajsutton replied:
I'd also suggest that this code to send subscribe and unsubscribe events should be split out into a separate class and then you can unit test it much more easily.

Something like:
```
public class CommitteeAssignmentManager {
  private final Map<Integer, Integer> committeeIndexToDeregisterSlot = new HashMap<>(); 
  private final EventBus eventBus; // set in constructor

  public void updateCommitteeAssignments(UnsignedLong slot, List<Integer> committeeIndices) {...}
}
```
The unit tests can then listen to the event bus to check that the right `CommitteeAssignmentEvent` and `CommitteeDismissalEvent` events are fired.

cemozerr replied:
Great ideas. I'm integrating them into the PR right now.

cemozerr replied:
Done.

ajsutton commented on networking/p2p/src/integration-test/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/GossipMessageHandlerIntegrationTest.java:
```diff
@@ -88,6 +105,116 @@ public void shouldGossipBlocksAcrossToIndirectlyConnectedPeers() throws Exceptio
+          assertThat(network2.getPeerManager().getAvailablePeerCount()).isEqualTo(1);
+        });
+
+    Thread.sleep(500);
```
Why do we need the sleeps?

cemozerr replied:
I explained in comments (in code) why I put sleeps.

ajsutton commented on networking/p2p/src/integration-test/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/GossipMessageHandlerIntegrationTest.java:
```diff
@@ -88,6 +105,116 @@ public void shouldGossipBlocksAcrossToIndirectlyConnectedPeers() throws Exceptio
+    chainUtil.initializeStorage(storageClient2);
+
+    // Setup VC 1
+    new ValidatorCoordinator(
```
I think actually creating the `ValidatorCoordinator` may be increasing the scope of the integration test too far here.  In my mind I'm kind of using the event bus to define the boundaries for integration tests.  So in this integration test I'd just fire `eventBus.new CommitteeAssignmentEvent())` to get the network layer to join subnets and fire `CommitteeDismissalEvent` to leave them again and check it works by checking if `Attestation` events and seeing whether or not they broadcast.

We might have a separate integration test for `ValidatorCoordinator` to check that it generates the right events given specific inputs (though I think we can cover that with unit tests if the committee assignment/dismissal events are managed by a separate class as mentioned below).

Otherwise we're rapidly reaching the scope of an acceptance test and the test gets brittle because we're having to build up multiple components and hook them together just right without using the actual production code to do that.

And conceptually, components which only interact via the event bus should be very loosely coupled so should be primarily tested independently (and we can check it all hooks together with a few acceptance tests to check everything actually works together but that's outside the scope of this work).

cemozerr committed:
Create CommiteeAssignmentManager

CommiteeAssignmentManager holds the commiteeAssignments map and
updates the assignments for the NEXT epoch, at the start of every
epoch. Also, it handles which commitee indices to register and deregister
(on the networking layer) by firing CommiteeAssignmentEvent and
CommiteeDismissalEvent instances.

cemozerr committed:
Add register/deregister tests

ajsutton commented on networking/p2p/src/integration-test/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/GossipMessageHandlerIntegrationTest.java:
```diff
@@ -88,6 +96,158 @@ public void shouldGossipBlocksAcrossToIndirectlyConnectedPeers() throws Exceptio
+        new CommitteeAssignmentEvent(List.of(validAttestation.getData().getIndex().intValue())));
+
+    // Sleep here because registering to new topics take time
+    Thread.sleep(1000);
```
ah right - it's awful that we need this wait but currently unavoidable as far as I know.  This is the same reason we have the sleep in the existing tests after first connecting the networks.  Could we extract a private method to encapsulate the sleep?  The tests will read a lot better that way and make the code self-documenting (ie `waitForTopicRegistration()` vs `Thread.sleep(1000)`).

Plus when we find a way to deterministically wait for subscriptions we can then just put it in one place.

cemozerr replied:
> Could we extract a private method to encapsulate the sleep?

I'm a bit confused about what you mean here.

ajsutton replied:
I'd introduce a private method:
```
private void waitForTopicRegistration() {
  Thread.sleep(1000);
}
```

It doesn't do anything different but makes the code much more readable but having a meaningful name.

cemozerr replied:
done.

ajsutton commented on networking/p2p/src/integration-test/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/GossipMessageHandlerIntegrationTest.java:
```diff
@@ -88,6 +96,158 @@ public void shouldGossipBlocksAcrossToIndirectlyConnectedPeers() throws Exceptio
+
+    Thread.sleep(1000);
+    // Sleep here because receiving new messages over the network takes time
+    assertTrue(network2Attestations.getAttestations().contains(validAttestation));
```
This sleep should be replaced with a `Waiter.waitFor(() -> assertTrue(...))` call.  That way the test progresses as soon as the message is received.

And it would be better to use the assertj `assertThat(network2Attestations.getAttestations()).contains(validAttestation)` as you'll get much more informative error messages.

cemozerr replied:
done.

ajsutton commented on validator/coordinator/src/main/java/tech/pegasys/artemis/validator/coordinator/CommitteeAssignmentManager.java:
```diff
@@ -0,0 +1,235 @@
+import tech.pegasys.artemis.util.bls.BLSSignature;
+import tech.pegasys.artemis.util.hashtree.HashTreeUtil;
+
+public class CommitteeAssignmentManager {
```
I think we can simplify this class down a bit by having it be "owned" by `ValidatorCoordinator`.  So rather than this class subscribing to events from the event bus `ValidatorCoordinator` holds an instance of it and is responsible for calling into it when assignments need to be calculated for each epoch.

I think the only public method this class then needs to expose is something like `public void updateCommitteeAssignments(final UnsignedLong epoch, final BeaconState state)` and that method would then fire `CommitteeAssignmentEvent` and `CommitteeDismissalEvent` events to the event bus as required.

By passing in the state we shouldn't need `CommitteeAssignmentManager` to have a `ChainStorageClient` which will make testing easier as there's less stuff to set up.

cemozerr replied:
done.

ajsutton commented on networking/p2p/src/integration-test/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/GossipMessageHandlerIntegrationTest.java:
```diff
@@ -88,6 +96,158 @@ public void shouldGossipBlocksAcrossToIndirectlyConnectedPeers() throws Exceptio
+    Thread.sleep(1000);
+    // Sleep here because receiving new messages over the network takes time
+
+    assertThat(network2AttestationsAfterDeregistration.getAttestations()).isEmpty();
```
Would be good to wrap this up in a `Waiter` method.  This is effectively a negative wait condition (ensuring that something doesn't happen) which will be fairly common.  So something in `Waiter` like:
```
public static void ensureConditionRemainsMet(final Condition condition)
      throws InterruptedException {
    final long mustBeTrueUntil = System.currentTimeMillis() + 2000;
    while (System.currentTimeMillis() < mustBeTrueUntil) {
      try {
        condition.run();
      } catch (final Throwable t) {
        throw new RuntimeException("Condition did not remain met", t);
      }
      Thread.sleep(500);
    }
  }
```
It could just sleep for 2 seconds than check but it's kind of nice to fail sooner if we know we've already received a message.

cemozerr replied:
done.

cemozerr committed:
Make VC own CommitteeAssignmentManager

cemozerr committed:
Implement CommitteeAssignmentManager tests

cemozerr committed:
Merge master into PR

ajsutton reviewed (approved):
LGTM.

cemozerr committed:
Encapsulate sleeping in wait methods

cemozerr committed:
Merge master into branch

cemozerr committed:
Fix merge introduced bugs
